File: packages/sdks/server-sdks/gatrix-dotnet-server-sdk/src/Gatrix.Server.Sdk/Cache/FlagDefinitionCache.cs
using Gatrix.Server.Sdk.Models;

namespace Gatrix.Server.Sdk.Cache;

/// <summary>
/// Thread-safe in-memory cache for feature flag definitions and segments.
/// Supports per-environment storage for multi-environment mode.
/// Environment must always be resolved by the caller (FeatureFlagService.ResolveEnvironment).
/// </summary>
public class FlagDefinitionCache
{
    private readonly object _lock = new();

    // Per-environment flag cache
    private readonly Dictionary<string, Dictionary<string, FeatureFlag>> _flagsByEnv = new(StringComparer.OrdinalIgnoreCase);

    // Segments are global (shared across environments)
    private Dictionary<string, FeatureSegment> _segments = new(StringComparer.OrdinalIgnoreCase);

    /// <summary>Get a flag definition by name for a specific environment.</summary>
    public FeatureFlag? GetFlag(string flagName, string environment)
    {
        lock (_lock)
        {
            if (_flagsByEnv.TryGetValue(environment, out var flags))
                return flags.GetValueOrDefault(flagName);

            return null;
        }
    }

    /// <summary>Get all cached flags for a specific environment (readonly snapshot).</summary>
    public List<FeatureFlag> GetCached(string environment)
    {
        lock (_lock)
        {
            if (_flagsByEnv.TryGetValue(environment, out var flags))
                return flags.Values.ToList();

            return [];
        }
    }

    /// <summary>Get all cached segments (readonly snapshot).</summary>
    public IReadOnlyDictionary<string, FeatureSegment> GetSegments()
    {
        lock (_lock)
        {
            return new Dictionary<string, FeatureSegment>(_segments, StringComparer.OrdinalIgnoreCase);
        }
    }

    /// <summary>Replace entire cache atomically with new definitions and segments for a specific environment.</summary>
    public void Update(IEnumerable<FeatureFlag> flags, IEnumerable<FeatureSegment> segments, string environment)
    {
        var newFlags = new Dictionary<string, FeatureFlag>(StringComparer.OrdinalIgnoreCase);
        foreach (var flag in flags)
            newFlags[flag.Name] = flag;

        var newSegments = new Dictionary<string, FeatureSegment>(StringComparer.OrdinalIgnoreCase);
        foreach (var segment in segments)
            newSegments[segment.Name] = segment;

        lock (_lock)
        {
            _flagsByEnv[environment] = newFlags;
            // Segments are merged globally (all environments share segments)
            foreach (var kvp in newSegments)
                _segments[kvp.Key] = kvp.Value;
        }
    }

    /// <summary>Upsert a single flag in a specific environment.</summary>
    public void UpsertFlag(FeatureFlag flag, string environment)
    {
        lock (_lock)
        {
            if (!_flagsByEnv.TryGetValue(environment, out var flags))
            {
                flags = new Dictionary<string, FeatureFlag>(StringComparer.OrdinalIgnoreCase);
                _flagsByEnv[environment] = flags;
            }
            flags[flag.Name] = flag;
        }
    }

    /// <summary>Remove a single flag by name from a specific environment.</summary>
    public void RemoveFlag(string flagName, string environment)
    {
        lock (_lock)
        {
            if (_flagsByEnv.TryGetValue(environment, out var flags))
                flags.Remove(flagName);
        }
    }

    /// <summary>Upsert a single segment (global, not per-environment).</summary>
    public void UpsertSegment(FeatureSegment segment)
    {
        lock (_lock)
        {
            _segments[segment.Name] = segment;
        }
    }

    /// <summary>Remove a single segment by name.</summary>
    public void RemoveSegment(string segmentName)
    {
        lock (_lock)
        {
            _segments.Remove(segmentName);
        }
    }

    public int FlagCount
    {
        get
        {
            lock (_lock)
            {
                return _flagsByEnv.Values.Sum(f => f.Count);
            }
        }
    }

    public void Clear()
    {
        lock (_lock)
        {
            _flagsByEnv.Clear();
            _segments.Clear();
        }
    }
}

File: packages/sdks/server-sdks/gatrix-java-server-sdk/src/main/kotlin/com/gatrix/server/sdk/cache/FlagDefinitionCache.kt
package com.gatrix.server.sdk.cache

import com.gatrix.server.sdk.models.FeatureFlag
import com.gatrix.server.sdk.models.FeatureSegment
import java.util.concurrent.ConcurrentHashMap

/**
 * Thread-safe in-memory cache for feature flag definitions and segments.
 * Supports per-environment flag storage and per-project segment storage.
 *
 * Segments are per-project (not global) — each project has its own segment namespace.
 * Environment-to-project mapping is maintained to resolve the correct segments.
 */
class FlagDefinitionCache {

    // Per-environment flag cache: envId -> (flagName -> FeatureFlag)
    private val flagsByEnv = ConcurrentHashMap<String, ConcurrentHashMap<String, FeatureFlag>>()

    // Per-project segment cache: projectId -> (segmentName -> FeatureSegment)
    private val segmentsByProject = ConcurrentHashMap<String, ConcurrentHashMap<String, FeatureSegment>>()

    // Environment-to-project mapping: environmentId -> projectId
    private val envToProjectMap = ConcurrentHashMap<String, String>()

    /**
     * Get a flag definition by name for a specific environment
     */
    fun getFlag(flagName: String, environmentId: String): FeatureFlag? {
        return flagsByEnv[environmentId]?.get(flagName)
    }

    /**
     * Get all cached flags for a specific environment (snapshot)
     */
    fun getCached(environmentId: String): List<FeatureFlag> {
        return flagsByEnv[environmentId]?.values?.toList() ?: emptyList()
    }

    /**
     * Get segments for a specific project
     */
    fun getSegments(projectId: String): Map<String, FeatureSegment> {
        return segmentsByProject[projectId]?.let { HashMap(it) } ?: emptyMap()
    }

    /**
     * Get segments for a specific environment (resolves via env->project mapping)
     */
    fun getSegmentsForEnvironment(environmentId: String): Map<String, FeatureSegment> {
        val projectId = envToProjectMap[environmentId] ?: return emptyMap()
        return getSegments(projectId)
    }

    /**
     * Get all segments merged across all projects (fallback for backward compat)
     */
    fun getAllSegments(): Map<String, FeatureSegment> {
        val merged = HashMap<String, FeatureSegment>()
        for (segmentMap in segmentsByProject.values) {
            merged.putAll(segmentMap)
        }
        return merged
    }

    /**
     * Replace all flags atomically for a specific environment
     * and update segments for the associated project
     */
    fun update(
        flags: List<FeatureFlag>,
        newSegments: List<FeatureSegment>,
        environmentId: String,
        projectId: String? = null
    ) {
        val newFlags = ConcurrentHashMap<String, FeatureFlag>()
        for (flag in flags) {
            newFlags[flag.name] = flag
        }
        flagsByEnv[environmentId] = newFlags

        // Track environment-to-project mapping and cache segments per project
        if (projectId != null) {
            envToProjectMap[environmentId] = projectId

            val projectSegments = segmentsByProject.computeIfAbsent(projectId) { ConcurrentHashMap() }
            // Replace all segments for this project
            projectSegments.clear()
            for (segment in newSegments) {
                projectSegments[segment.name] = segment
            }
        }
    }

    /**
     * Upsert a single flag in a specific environment
     */
    fun upsertFlag(flag: FeatureFlag, environmentId: String) {
        val flags = flagsByEnv.computeIfAbsent(environmentId) { ConcurrentHashMap() }
        flags[flag.name] = flag
    }

    /**
     * Remove a single flag by name from a specific environment
     */
    fun removeFlag(flagName: String, environmentId: String) {
        flagsByEnv[environmentId]?.remove(flagName)
    }

    /**
     * Upsert a single segment for a specific project
     */
    fun upsertSegment(segment: FeatureSegment, projectId: String) {
        val segments = segmentsByProject.computeIfAbsent(projectId) { ConcurrentHashMap() }
        segments[segment.name] = segment
    }

    /**
     * Remove a single segment by name from a specific project
     */
    fun removeSegment(segmentName: String, projectId: String) {
        segmentsByProject[projectId]?.remove(segmentName)
    }

    /**
     * Remove a segment from all projects (when projectId is unknown)
     */
    fun removeSegmentFromAll(segmentName: String) {
        for (segmentMap in segmentsByProject.values) {
            segmentMap.remove(segmentName)
        }
    }

    /**
     * Get project ID for a given environment
     */
    fun getProjectId(environmentId: String): String? {
        return envToProjectMap[environmentId]
    }

    /**
     * Total number of flags across all environments
     */
    val flagCount: Int
        get() = flagsByEnv.values.sumOf { it.size }

    /**
     * Get list of cached environment IDs
     */
    fun getEnvironmentIds(): List<String> {
        return flagsByEnv.keys().toList()
    }

    /**
     * Clear all cached data
     */
    fun clear() {
        flagsByEnv.clear()
        segmentsByProject.clear()
        envToProjectMap.clear()
    }
}
